### ISSUE TITLE
Feature request: Render the text of the references inline

### ISSUE THREAD

**Issue by danrthompson** [TfTHacker/obsidian42-strange-new-worlds#7]
First of all, I want to thank you so much for making this plugin!!

Honestly, this is one of the best plugins I have (and I’ve tried out over one hundred). In fact, I think the devs should incorporate this functionality natively because it is so important to the goal of surfacing connections.

Also, I saw you were involved with Roam42, which means you made the most useful plugins I had when I used Roam. And then you made this plugin, which helped convince me to switch from Roam to obsidian! So thanks very much, I really appreciate your work :).

Anyway, on to my actual message.

I'm sure you've thought of this already, but it would be amazing to be able to see the text of the references inline.

This plugin gets me halfway there in terms of surfacing new connections and helping me to find / revisit helpful info. But when the list merely shows the file name, I get some info from that, but to really see what the connection is about I need to read the block. And to do that, I have to click on each link individually, which is tedious. And so in practice I end up not doing that.

This problem is exacerbated when the reference is from a daily note. Because in that case the file name is merely a date, which provides very little context as to what the connecting idea actually is.

I know that adding the text of each reference could potentially clutter things, but that can be mitigated. For example, it could potentially be set up to only show the references' text when the user clicks on a button. Perhaps one button per reference if you want to toggle it for just that reference, and also an "expand / collapse all" button at the top.

Thank you so much, again, for all the amazing work. I can't wait to see where this plugin goes!

**Comment by TfTHacker** [comment]
I completely agree with you. I will see if I can solve this, but the issue is that Obsidian treats embeds as something unique. I am passing the block from the page to the standard "markdown rendering" API, however, Obsidian returns everything but the block embeds as rendered text.

But I agree and will keep an eye on how to solve this, if possible.

Very thankful for your writing this detailed description and also beta testing. Keep the feedback coming. This plugin is the hardest I have ever developed. It looks simple on the surface, but obsidian's internals for this kind of thing is hard to navigate.

**Comment by danrthompson** [comment]
Of course! I think this plugin is honestly one of the best I've used. I also find it to be extremely fast and completely stable. I think there are a ton of people who would find it to be really useful, so I personally would advocate releasing it through the community plugin marketplace. I think it would not only be super useful for others, but would also show people how powerful some of the "Roam-like" features are. I've noticed that there seems to be a contingent of people who perhaps are more used to traditional markdown or notetaking apps, and who seem to object to most block-based or metadata based features, but I suspect they've just never seen something like this.

In terms of stability, I actively use over 50 plugins, and have not had any interference or other issues. So I really don't see a downside to wider release.

Of course that's just my humble opinion, and more than anything I'm just personally thrilled this exists!!

**Comment by danrthompson** [comment]
Separately, it looks like you did this with a hover view!! And it is awesome!

Two small thoughts regarding this feature.

1. I would enlarge the hover popup and reduce the text size so you can see more links and more context.
2. This is more of a feature request, and I'm sure it's not trivial at all. But it would be incredible to be able to filter the references using something like the search feature in the backlinks pane. I have a lot of thoughts about what this search would ideally look like, and I can share those if you're interested. But the tldr version is that any search would be great, but ideally it would be possible to search based on
    1. File path
    2. Page metadata (like tags, links, text etc anywhere on the page of the linking reference).
    3. Block text and tags (tags and text present in the block that made the link)
3. Also more of a feature request, but it would be awesome to be able to see more context for each reference. Basically, making it more like the backlinks pane would be fantastic. I wonder if there is a way to access a view like that through the API itself?

Let me know if you'd like me to break these out into separate issues/feature requests, btw.

Finally, I don't mean to be demanding by providing all this feedback, just mentioning it since you asked :)

Overall, I really think this is such an incredible plugin, and thanks again for making it. It really is incredible.

**Comment by danrthompson** [comment]
On a personal note, I'm curious about your opinion about Roam vs Obsidian. I know this isn't exactly the ideal forum, but not sure where else I'd ask you.

If you worked on Roam42, you obviously have a ton of experience with both.

I started with Roam, and eventually migrated to Obsidian because of the plugin ecosystem and dataview in particular. Plus the userbase and pace of development just seemed to be stronger and improving (whereas Roam's seemed to be declining).

My biggest concern with Obsidian however is that it seems like:
1. The devs are not inclined to replicate the markdown file contents in a database
2. They also don't seem super inclined to enhance "block" oriented features (like better transclusion, block references like in this plugin, etc)
3. The first point means that, despite the existence of Dataview, it's hard to really query and modify the notes programmatically or structurally or using code because the index is limited and the ground truth is not in a structured database, but rather in messy markdown files

I suppose overall there is a level of polish and even extensibility (ironically, because Obsidian's plugin ecosystem is SO much better, but nonetheless it seems that the database infrastructure that Roam has makes more things possible) that Roam has and that I assumed Obsidian would replicate eventually. But now I'm not so sure.

Also frankly I'm concerned about funding, although with 2 devs it seems fairly sustainable. But overall they'll just have FAR fewer resources than Roam when they offer the main product for free.

Of course, there are plenty of wonderful things about Obsidian, and its text-based nature. But those are some of my concerns. 

But basically, just curious to get your opinion. As a plugin developer for both, which tool do you think has the potential to be more powerful? Is Roam's plugin system likely to improve? Which tool do you prefer and why?

No pressure to answer any of this btw, just curious!

**Comment by TfTHacker** [comment]
First, thanks for the praise. We need it. We have put days and days into this and of course its a hobby project. It is one of those things if you knew the upfront cost you wouldn't do, but once you are so deep, there is no turning around. We will release to broad testing audience soon. But we have a few more "low hanging fruit" bugs to fix.

Regarding your ideas, they are great. would you mind submitting individual issues as feature request. This will help us keep up a dialogue and I can track the FR to completion. Love the idea of integrated search. Also been playing with the idea of the size of the hover control. FYI: we intentionally made everything very CSS controllable, so you can really tweak it to your own needs. But it might be nice to put together some working CSS examples that create different effects.

Regarding Roam and Obisidan. All your concerns are valid, but I think things have improved. Obsidian has hired a few devs and already in last month we are seeing a faster rate of updates/fixes. I had the same concern, but now I don't.  Regarding db to markdown, I know the dataview developer is planning in the near future to make dataview queries render to markdown so that people are not dependent on dataview plugin if they want to export content or move away.  Not sure if this addresses your concern.

Feel free to ping me as well on Discord. Really appreciate your testing SNW.

**Comment by danrthompson** [comment]
Ah specifically when I mentioned the advantage of a database I was not referring to that making it more or less exportable. I absolutely love dataview and use it all the time, and it absolutely would cover that use case. Not to mention that obsidian is naturally exported since it just lives in markdown files.

What I meant by Roam maybe getting an advantage from being based upon a database is that it likely makes it far easier to enable certain features (either natively or through plugins). For example, better outlining tools, plugins like Smartblocks, native block references, and editable transclusions are certainly theoretically possible in obsidian. And some of those features will probably eventually be added. But it seems to be a lot harder when you're working with markdown files instead of structured data in a database. At least I assume so - I haven't actually gone through the Obsidian API, so I don't know what kind of data structures they give you.

So I guess I should stop here and ask you since you're obviously familiar with the Obsidian API and whatever indexing they do:

Is it harder to read and write note data in obsidian? What about for queries that treat the notes as blocks in an outline (or graph)? As in, not in page level units but rather in block level units (each paragraph, list item, etc).

As I understand it, at least some info doesn't seem to be available in the index. Like the text of a block, except for lists and tasks. Or at least dataview doesn't expose it. And I assume whatever index is available doesn't enable writes. 

Maybe I'm assuming incorrectly though? Like if there is an abstraction layer over the notes that basically stores them in a data structure that enables reads and also writes that are synced to the note?

**Comment by danrthompson** [comment]
So two things:
1. I'm not surprised to hear that markdown is more complex to work with than a DB. I guess I was hoping that the Ovsidian API exposed an abstraction layer that handled retrieving the markdown contents with a structured API, handling the edge cases internally. And I thought it unlikely but possible that maybe they even exposed that same abstraction for writes to make it completely seamless for devs. If they did that, I think it would eliminate the disadvantage of not having a DB, except possibly performance considerations.

But sounds like this is not something they have right now? If not that's a little unfortunate. How much harder have you found it to develop for obsidian vs roam?

I also wonder if it would be worth pushing for a data structure or abstraction later like the one I just described. If it would simplify development, it seems like it would really neatly remove all the complexity of dealing with markdown, making file system storage exclusively a benefit. 

Finally, it seemed like you closed this ticket saying that you weren't going to tackle it at the moment. But it's already done isn't it? I added two other feature requests but this one was about going from the previous UI that just showed a file and line number for each reference to one that showed the actual text itself for each reference. And that appears to be implemented.

Did you interpret this FR differently?

**Comment by TfTHacker** [comment]
Regarding this incident, yes its resolved it open since we had a few other things going on in our conversation. 

Obsidian has an API that does keep a current cache of most of the interesting things in the markdown, but not everything. It isn't easy to pull from the API as well. The API isn't as easy to work with as with other api's. Roam was easier for some things, but also really hard for some things.

The Obsidian API is actually great to work with for most things, just this block ref count is a unique take on working with markdown. For this a real database would be much easier to work with.

Another challenge of Obsidian is there is an editor and a preview mode, which are both different technologies. So for the counter, we have to repeat the logic twice and in a different format.

I really like your other recommendations for this plugin. I am thinking about how to incorporate them. One issue that comes up in my mind is to avoid repeating what is built in. For example, I can create expanding previews of inline block references. However the hover control with a mouse over the preview renders a hover window with the entire document that can be scrolled. So this is the desired functionality, but chaining together a few features to get to the results. The amount of code and testing verses using something that already we have is what I weigh out. 

I would rather spend my time I think on unique features of SNW and tap into existing features.

**Comment by danrthompson** [comment]
All makes sense. I hope the obsidian developers prioritize making the API even more convenient and robust, since so much of the functionality is from plugins.

And absolutely no pressure at all to implement any of the FR's if they are too tough or just not your priority. I wanted to give you my feedback since you've made this awesome tool. But I will be thrilled to see any functionality you deliver. And in fact if you have ideas for things that are novel, that sounds awesome!

I guess one small thought, and again, please ignore if it's not useful. But while some of the functionality in the FR's I opened already exists (and I agree is lower priority as a result), it currently isn't easy to see inbound references for things other than pages. Just to throw that out there

But anyway can't wait to see what you end up making!

If you have a backlog of feature ideas I'd be super curious to see it btw. Not in the sense of wondering when things will be implemented but rather to just see what ideas you have. I could see this going in a few interesting directions. I guess my overall conception for it is that it makes it easier to get more "peripheral vision" around the notes. 

In the sense that when reading a text file you have no connection to other notes. In a hypertext document you see outbound links one hop away. In obsidian without plugins you have one hop in each direction in the note. Plus the graph view, which adds quite a bit more, but which also has its own limitations (no context for where the links are coming from in particular).

This plugin adds additional peripheral information in two ways:
1. It shows one hop links to more entities within the note. Yes they would've been in backlinks but seeing the links while reading the note, and also seeing where specifically they go, is useful.
2. The numbers show some indication about which sections, notes, etc are highly cited.
3. Finally, and most uniquely, in my opinion, it shows you notes more than one hop away by showing you notes that also linked to the same note or section that you linked to. So you see notes that are indirectly connected. Which I think is a really interesting concept and one that is surprisingly unexplored.

One other plugin also surfaces these indirectly connected notes, but it is super opinionated and requires extensive use of tags in a particular way, which makes it not useful for me at the moment. But it did make me think, so that alone was cool. Anyway in case you find it interesting yourself, the plugin is called "Tag Navigator," formerly "Cross reference navigation."

As an aside, I've been wondering about how to go further with this theme. One thought I have is to feed all of my notes into OpenAI's GPT-3 so that I can then use natural language semantic queries to surface connections that I'd likely miss otherwise. But have not yet gotten around to that. I'm actually shocked that this is not really a thing yet.
